message/infrastructure/googlecloud/subscriber.go

```diff
@@ -0,0 +1,221 @@
+package googlecloud
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"sync"
+
+	"google.golang.org/api/option"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/pkg/errors"
+)
+
+var (
+	ErrSubscriberClosed = errors.New("subscriber is closed")
+)
+
+type subscriber struct {
+	closing chan struct{}
+	closed  bool
+
+	allSubscriptionsWaitGroup sync.WaitGroup
+	activeSubscriptions       map[string]struct{}
+	activeSubscriptionsLock   sync.RWMutex
+
+	client *pubsub.Client
+	config SubscriberConfig
+
+	unmarshaler Unmarshaler
+	logger      watermill.LoggerAdapter
+}
+
+type SubscriberConfig struct {
+	SubscriptionName            string
+	ProjectID                   string
+	CreateSubscriptionIfMissing bool
```

Comment:
also, sounds like redundant (we always want to do this?)

Reply:
I wonder which case is more frequent: 
1) I am calling `Subscribe` and don't want to care whether there is a subscription or not,
or
2) I only want to use preexisting subscriptions and receive an error if no such subscription is present.

I suppose 1) is way more prevalent. I cannot think of a case where I would want 2), but I guess it's possible. This could be left as a configuration detail, but:

a) Like you say, the default would probably need to be 1), as this seems to be the more natural behavior,
b) It may be difficult to convey this configuration option clearly, so that it's self-explanatory and doesn't pollute the config.

Reply:
lgtm